price-feeder/telemetry/metrics.go

```diff
@@ -0,0 +1,173 @@
+	EnableHostname bool `toml:"enable_hostname"`
+
+	// Enable adding hostname to labels
+	EnableHostnameLabel bool `toml:"enable_hostname_label"`
+
+	// Enable adding service to labels
+	EnableServiceLabel bool `toml:"enable_service_label"`
+
+	// PrometheusRetentionTime, when positive, enables a Prometheus metrics sink.
+	// It defines the retention duration in seconds.
+	PrometheusRetentionTime int64 `toml:"prometheus_retention_time"`
+
+	// GlobalLabels defines a global set of name/value label tuples applied to all
+	// metrics emitted using the wrapper functions defined in telemetry package.
+	//
+	// Example:
+	// [["chain_id", "cosmoshub-1"]]
+	GlobalLabels [][]string `toml:"global_labels"`
+}
+
+// Metrics defines a wrapper around application telemetry functionality. It allows
+// metrics to be gathered at any point in time. When creating a Metrics object,
+// internally, a global metrics is registered with a set of sinks as configured
+// by the operator. In addition to the sinks, when a process gets a SIGUSR1, a
+// dump of formatted recent metrics will be sent to STDERR.
+type Metrics struct {
+	memSink           *metrics.InmemSink
+	prometheusEnabled bool
+}
+
+// GatherResponse is the response type of registered metrics
+type GatherResponse struct {
+	Metrics     []byte
+	ContentType string
+}
+
+// New creates a new instance of Metrics
+func New(cfg Config) (*Metrics, error) {
+	if !cfg.Enabled {
+		return nil, nil
+	}
+
+	if numGlobalLables := len(cfg.GlobalLabels); numGlobalLables > 0 {
+		parsedGlobalLabels := make([]metrics.Label, numGlobalLables)
+		for i, gl := range cfg.GlobalLabels {
+			parsedGlobalLabels[i] = NewLabel(gl[0], gl[1])
+		}
+
+		globalLabels = parsedGlobalLabels
+	}
+
+	metricsConf := metrics.DefaultConfig(cfg.ServiceName)
+	metricsConf.EnableHostname = cfg.EnableHostname
+	metricsConf.EnableHostnameLabel = cfg.EnableHostnameLabel
+
+	memSink := metrics.NewInmemSink(10*time.Second, time.Minute)
+	metrics.DefaultInmemSignal(memSink)
+
+	m := &Metrics{memSink: memSink}
+	fanout := metrics.FanoutSink{memSink}
+
+	if cfg.PrometheusRetentionTime > 0 {
+		m.prometheusEnabled = true
+		prometheusOpts := metricsprom.PrometheusOpts{
+			Expiration: time.Duration(cfg.PrometheusRetentionTime) * time.Second,
+		}
+
+		promSink, err := metricsprom.NewPrometheusSinkFrom(prometheusOpts)
+		if err != nil {
+			return nil, err
+		}
+
+		fanout = append(fanout, promSink)
+	}
+
+	if _, err := metrics.NewGlobal(metricsConf, fanout); err != nil {
+		return nil, err
+	}
+
+	return m, nil
+}
+
+// Gather collects all registered metrics and returns a GatherResponse where the
+// metrics are encoded depending on the type. Metrics are either encoded via
+// Prometheus or JSON if in-memory.
+func (m *Metrics) Gather(format string) (GatherResponse, error) {
+	switch format {
+	case FormatPrometheus:
+		return m.gatherPrometheus()
+
+	case FormatText:
+		return m.gatherGeneric()
+
+	case FormatDefault:
```

Comment:
I think we can combine these? `case FormatText, FormatDefault:`. Give that a shot.